# Рубежный контроль № 1
# По дисциплине:
### Методы машинного обучения в АСОИУ
____
Выполнил:  
Сутдент гр. ИУ5-22М  
Бумагин П. В.

**Вариант 1**
## Задание
### **Задача № 1**
Для набора данных проведите кодирование одного (произвольного) категориального признака с использованием метода "count (frequency) encoding".
### **Задача № 21**
Для набора данных проведите масштабирование данных для одного (произвольного) числового признака с использованием масштабирования по медиане.
### **Дополнительное задание:**  
для произвольной колонки данных построить гистограмму


### **Загрузим данные датасета**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import RobustScaler

# Настройка стиля графиков
sns.set_theme(style="whitegrid")

# Загружаем датасет напрямую по URL-адресу из репозитория seaborn-data
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
df = pd.read_csv(url)

print(f"Размерность набора данных: {df.shape}")
print("\nПервые 5 строк датасета:")
print(df.head())

Matplotlib is building the font cache; this may take a moment.


Размерность набора данных: (244, 7)

Первые 5 строк датасета:
   total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4


### Описание набора данных «Tips» (Чаевые)

Для выполнения работы выбран классический набор данных **«Tips»**, который часто используется в качестве демонстрационного в библиотеке визуализации данных `Seaborn`. Исторически эти данные были собраны официантом одного из ресторанов США в течение нескольких месяцев работы. 

#### 1. Основные характеристики датасета:
* **Количество записей (строк):** 244
* **Количество признаков (столбцов):** 7
* **Наличие пропусков:** Данные полностью заполнены (пропуски отсутствуют).

#### 2. Структура признаков:

| Имя признака | Тип данных | Описание | Тип шкалы |
| :--- | :--- | :--- | :--- |
| `total_bill` | Вещественный (float) | Общая сумма счета за заказ в долларах США | Количественный непрерывный |
| `tip` | Вещественный (float) | Размер чаевых в долларах США | Количественный непрерывный |
| `sex` | Категориальный (binary) | Пол клиента, оплатившего счет (`Male` / `Female`) | Качественный номинальный |
| `smoker` | Категориальный (binary) | Наличие курильщиков в компании клиентов (`Yes` / `No`) | Качественный номинальный |
| `day` | Категориальный (4 градации) | День недели (`Thur`, `Fri`, `Sat`, `Sun`) | Качественный номинальный |
| `time` | Категориальный (binary) | Время приема пищи (`Lunch` — обед / `Dinner` — ужин) | Качественный номинальный |
| `size` | Целочисленный (int) | Размер компании клиентов (количество человек за столом) | Количественный дискретный |

#### 3. Обоснование выбора признаков для выполнения задач:

* **Для Задачи №1 (Count / Frequency Encoding):** 
  Признак `day` (день недели) является категориальным и принимает 4 различных значения. Он хорошо подходит для демонстрации частотного кодирования, так как разные дни недели встречаются в выборке с разной регулярностью (например, суббота и воскресенье посещаются чаще, чем четверг и пятница).
  
* **Для Задачи №21 (Масштабирование по медиане):** 
  Признак `total_bill` (общий счет) представляет собой непрерывную числовую величину. Он имеет выраженное смещение вправо (асимметрию) и несколько крупных единичных значений (счетов более $40), которые можно классифицировать как выбросы. Это позволяет наглядно продемонстрировать преимущества масштабирования по медиане и межквартильному размаху (Robust Scaling), которое, в отличие от MinMaxScaler или StandardScaler, не сжимает основную массу данных в узкий диапазон под влиянием аномальных значений.